ensembl_dbcopy/api/views.py

```diff
@@ -11,75 +11,120 @@
    See the License for the specific language governing permissions and
    limitations under the License.
 """
-from mysql.connector import (connection)
+import re
+from functools import lru_cache
+import sqlalchemy as sa
 from rest_framework.response import Response
 from rest_framework.views import APIView
+from rest_framework import status
 
-from ensembl_dbcopy.models import Host
+from ensembl_dbcopy.models import Host, Dbs2Exclude
 import logging
 
+
 logger = logging.getLogger(__name__)
 
-def mysql_connection(query_params=None):
-    # host = query_params.get('host', None)
-    if query_params is None:
-        query_params = {}
-    logger.debug('Query Params %s', query_params)
-    host = Host.objects.filter(name=query_params.get('host', 'localhost'),
-                               port=query_params.get('port', 3306)).first()
-    password = query_params.get('password', None)
-    if host:
-        logger.debug('Retrieved host %s', host)
-        return connection.MySQLConnection(user=host.mysql_user,
-                                          host=host.name,
-                                          port=host.port,
-                                          password=password,
-                                          database='information_schema')
-    raise RuntimeError('No host corresponding to parameters %s' % query_params)
+
+def make_excluded_schemas():
+    schemas = set()
+    def closure():
+        if not schemas:
+            schemas.update(Dbs2Exclude.objects.values_list('table_schema', flat=True))
+        return schemas
+    return closure
+
+get_excluded_schemas = make_excluded_schemas()
+
+
+@lru_cache(maxsize=None)
```

Comment:
What's the purpose of `lru_cache` here?

Reply:
To reuse the same engine object (and its connection pool) when querying the same host multiple times in a rapid succession (which is the use case for the autocomplete feature that queries the server at pretty much each keystroke). The connection pool is recycled every hour so they won't linger on the server. Also this saves queries to the host DB. The only downside is that if we change a host port or mysql username we'll have to reload gunicorn workers, but that will rarely happen I believe.

Reply:
thanks for the insight!